R/patch.R

```diff
@@ -0,0 +1,375 @@
+  }
+
+  # Extract metadata:
+  earlier_n <- nrow(earlier_snapshot)
+  later_n <- nrow(later_tbl)
+  tbl_names <- names(earlier_snapshot)
+  val_names <- tbl_names[!tbl_names %in% ukey_names]
+
+  # More input validation:
+  if (!identical(tbl_names, names(later_tbl))) {
+    cli_abort(c(
+      "`earlier_snapshot` and `later_tbl` should have identical column
+       names and ordering.",
+      "*" = "`earlier_snapshot` colnames: {format_chr_deparse(tbl_names)}",
+      "*" = "`later_tbl` colnames: {format_chr_deparse(names(later_tbl))}"
+    ), class = "epiprocess__tbl_diff2__tbl_names_differ")
+  }
+
+  combined_tbl <- vec_rbind(earlier_snapshot, later_tbl)
+  combined_n <- nrow(combined_tbl)
+
+  # We'll also need epikeytimes and value columns separately:
+  combined_ukeys <- combined_tbl[ukey_names]
+  combined_vals <- combined_tbl[val_names]
+
+  # We have five types of rows in combined_tbl:
+  # 1. From earlier_snapshot, no matching ukey in later_tbl (deletion; turn vals to
+  #    NAs to match epi_archive format)
+  # 2. From earlier_snapshot, with matching ukey in later_tbl (context; exclude from
+  #    result)
+  # 3. From later_tbl, with matching ukey in earlier_snapshot, with value "close" (change
+  #    that we'll compactify away)
+  # 4. From later_tbl, with matching ukey in earlier_snapshot, value not "close" (change
+  #    that we'll record)
+  # 5. From later_tbl, with no matching ukey in later_tbl (addition)
+
+  # For "snapshot" later_format, we need to filter to 1., 4., and 5., and alter
+  # values for 1.  For "update" later_format, we need to filter to 4. and 5.
+
+  # (For compactify_abs_tol = 0, we could potentially streamline things by dropping
+  # ukey+val duplicates (cases 2. and 3.).)
+
+  # Row indices of first occurrence of each ukey; will be the same as
+  # seq_len(combined_n) except for when that ukey has been re-reported in
+  # `later_tbl`, in which case (3. or 4.) it will point back to the row index of
+  # the same ukey in `earlier_snapshot`:
+  combined_ukey_firsts <- vec_duplicate_id(combined_ukeys)
+
+  # Which rows from combined are cases 3. or 4.?
+  combined_ukey_is_repeat <- combined_ukey_firsts != seq_len(combined_n)
+  # For each row in 3. or 4., row numbers of the ukey appearance in earlier:
+  ukey_repeat_first_i <- combined_ukey_firsts[combined_ukey_is_repeat]
+
+  # Which rows from combined are in case 3.?
+  combined_compactify_away <- rep(FALSE, combined_n)
+  combined_compactify_away[combined_ukey_is_repeat] <-
+    vec_approx_equal0(combined_vals,
+      combined_vals,
+      na_equal = TRUE,
+      abs_tol = compactify_abs_tol,
+      inds1 = combined_ukey_is_repeat,
+      inds2 = ukey_repeat_first_i
+    )
+
+  # Which rows from combined are in cases 3., 4., or 5.?
+  combined_from_later <- vec_rep_each(c(FALSE, TRUE), c(earlier_n, later_n))
+
+  if (later_format == "update") {
+    # Cases 4. and 5.:
+    combined_tbl <- combined_tbl[combined_from_later & !combined_compactify_away, ]
```

Comment:
Case 1 isn't relevant here because it's indistinguishable from unchanged rows when we're just updating, right?